pkg/k8s/engine_controller.go

```diff
@@ -149,6 +155,32 @@ func (c *EngineController) shouldHandleGateway(gateway *v1alpha1.Gateway) bool {
 	return false
 }
 
+func (c *EngineController) completeGateway(gateway *v1alpha1.Gateway) error {
+	if gateway.Status.ActiveEndpoint.NodeName != c.nodeName {
+		return nil
+	}
+	publicIP, err := utils.GetPublicIP()
+	if err != nil {
+		return err
+	}
+	// retry to update ip
+	err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		var i int
+		for i = 0; i < len(gateway.Spec.Endpoints); i++ {
```

Comment:
How about using `for range` here and putting the update logic inside the iteration. Like this:
```
        // retry to update ip
	err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
		for k, v := range gateway.Spec.Endpoints {
			if v.NodeName == c.nodeName {
				gateway.Spec.Endpoints[k].PublicIP = publicIP
				_, err = c.ravenClient.RavenV1alpha1().Gateways().Update(context.Background(), gateway, v1.UpdateOptions{})
				return err
			}
		}
		return nil
	})
```

Reply:
done